### PR TITLE
Bluetooth: Update advertising data without advertising restart

### PR DESCRIPTION
**Pull request by MariuszSkamra**
This adds new API fuction to update running advertising data.
It will remove the need of advertising restarting.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

### CODE REVIEW

File: subsys/bluetooth/host/hci_core.c [hunk 1]
```diff
@@ -5121,6 +5121,79 @@ static inline bool ad_has_name(const struct bt_data *ad, size_t ad_len)
...
+	return 0;
+}
+
+int bt_le_adv_update(const struct bt_data *ad, size_t ad_len,
+		     const struct bt_data *sd, size_t sd_len)
+{
+	bool connectable, use_name;
+
+	if (!atomic_test_bit(bt_dev.flags, BT_DEV_ADVERTISING)) {
+		return -ENODEV;
```
Comment (jhedberg) [inline subsys/bluetooth/host/hci_core.c]: Why ENODEV? Perhaps EAGAIN?

Comment (MariuszSkamra) [reply subsys/bluetooth/host/hci_core.c]: I couldn't find an error that could fit here TBH. Could be EAGAIN :)

Comment (MariuszSkamra) [reply subsys/bluetooth/host/hci_core.c]: done

**Comment by jhedberg** [comment]
One thing that this doesn't doesn't seem to address is multiple advertising sets (e.g. take the set to update as a parameter). Also is there anything to take into account regarding AE? We shouldn't add any new APIs until we're clear on these issues.

**Review by jhedberg — CHANGES_REQUESTED** [review]
Marking as "request changes" until we've discussed/resolved the AE impact

**Comment by sjanc** [comment]
I think AE should have separate API (mutually exclusive with legacy), eg with max 1650 bytes of payload flat buffers are no-go and API will have to use netbufs probably.

**Review by sjanc — APPROVED** [review]

**Comment by MariuszSkamra** [comment]
@jhedberg Looking into advertising API, it is not prepared for AE. I agree with @sjanc that AE should have separate API functions.

**Review by jhedberg — APPROVED** [review]

File: include/bluetooth/bluetooth.h [hunk 2]
```diff
@@ -348,6 +348,20 @@ int bt_le_adv_start(const struct bt_le_adv_param *param,
...
+ *  Update advertisement and scan response data.
+ *
+ *  @param ad Data to be used in advertisement packets.
+ *  @param ad_len Number of elements in ad
+ *  @param sd Data to be used in scan response packets.
+ *  @param sd_len Number of elements in sd
+ *
+ *  @return Zero on success or (negative) error code otherwise.
+ */
+int bt_le_adv_update(const struct bt_data *ad, size_t ad_len,
```
Comment (carlescufi) [inline include/bluetooth/bluetooth.h]: I would probably call this `bt_le_adv_update_data`

Comment (jhedberg) [reply include/bluetooth/bluetooth.h]: Good point - that would describe better what it does (e.g. it doesn't update the advertising parameters)

**Review by carlescufi — APPROVED** [review]

**Pull request merged**